roachtest: knex failed

roachtest.knex [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10044633?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RoachtestNightlyGceBazel/10044633?buildTab=artifacts#/knex) on master @ [87c9598a9d59433775764c5a78604c2d3a488ad1](https://github.com/cockroachdb/cockroach/commits/87c9598a9d59433775764c5a78604c2d3a488ad1):


```
test artifacts and logs in: /artifacts/knex/run_1
(knex.go:124).func1: COMMAND_PROBLEM: ssh verbose log retained in ssh_092938.134428279_n1_cd-mntdata1knex-DBco.log: exit status 3
```
<p>Parameters: <code>ROACHTEST_cloud=gce</code>
, <code>ROACHTEST_cpu=4</code>
, <code>ROACHTEST_encrypted=false</code>
, <code>ROACHTEST_ssd=0</code>
</p>
<details><summary>Help</summary>
<p>


See: [roachtest README](https://github.com/cockroachdb/cockroach/blob/master/pkg/cmd/roachtest/README.md)



See: [How To Investigate \(internal\)](https://cockroachlabs.atlassian.net/l/c/SSSBr8c7)

</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #103174 roachtest: knex failed [C-test-failure O-roachtest O-robot branch-release-22.1 release-blocker]
- #103172 roachtest: knex failed [C-test-failure O-roachtest O-robot T-sql-sessions branch-release-22.2 release-blocker]
</p>
</details>
/cc @cockroachdb/sql-sessions
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*knex.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

```
  1) CLI tests
       knexfile
         knexfile resolution
           --cwd is specified
             and --knexfile is also specified
               and --knexfile is an absolute path
                 uses the indicated knexfile:
     Error: Expected output to include "Batch 1 run: 4 migrations", but it was actually "Failed to load external module ts-node/register
Failed to load external module typescript-node/register
Failed to load external module typescript-register
Failed to load external module typescript-require
Failed to load external module sucrase/register/ts
Failed to load external module @babel/register
Working directory changed to /mnt/data1/knex/test/jake-util/knexfile-ts
"
      at new AssertionError (node_modules/cli-testlab/lib/AssertionError.ts:3:5)
      at assertOutput (node_modules/cli-testlab/lib/execution-test-helper.ts:71:14)
      at Exec.<anonymous> (node_modules/cli-testlab/lib/execution-test-helper.ts:130:9)
      at Exec.emit (node:events:513:28)
      at ChildProcess.<anonymous> (node_modules/cli-testlab/node_modules/jake/lib/utils/index.js:227:20)
      at ChildProcess.emit (node:events:513:28)
      at Process.ChildProcess._handle.onexit (node:internal/child_process:293:12)

  2) Integration Tests
       postgresql | cockroachdb
         Deletes
           Delete with join
             should handle basic delete with join:
     AssertionError: expected promise to be rejected but it was fulfilled with +0
  

  3) Integration Tests
       postgresql | cockroachdb
         Deletes
           Delete with join
             should handle returning:
     AssertionError: expected promise to be rejected but it was fulfilled with []
```

The test code already is expecting these errors, but looks like we need to update the test name parsing.

--- a/pkg/cmd/roachtest/tests/knex.go
+++ b/pkg/cmd/roachtest/tests/knex.go
@@ -22,7 +22,7 @@ import (
 	"github.com/stretchr/testify/require"
 )
 
-const supportedKnexTag = "2.0.0"
+const supportedKnexTag = "2.5.1"
 
 // This test runs one of knex's test suite against a single cockroach
 // node.

--- a/pkg/cmd/roachtest/tests/typeorm.go
+++ b/pkg/cmd/roachtest/tests/typeorm.go
@@ -14,6 +14,7 @@ import (
 	"context"
 	"fmt"
 	"regexp"
+	"strconv"
 	"strings"
 
 	"github.com/cockroachdb/cockroach/pkg/cmd/roachtest/cluster"
@@ -166,12 +167,27 @@ func registerTypeORM(r registry.Registry) {
 		rawResults := result.Stdout + result.Stderr
 		t.L().Printf("Test Results: %s", rawResults)
 		if err != nil {
+			// We don't have a good way of parsing test results from javascript, so we
+			// use substring matching and regexp instead of using a blocklist like
+			// we use for other ORM tests.
+			numFailingRegex := regexp.MustCompile(`(\d+) failing`)
+			matches := numFailingRegex.FindStringSubmatch(rawResults)
+			numFailing, convErr := strconv.Atoi(matches[1])
+			if convErr != nil {
+				t.Fatal(convErr)
+			}
+
+			// One test is known to flake during setup.
+			if strings.Contains(rawResults, `"before each" hook for "should select specific columns":`) {
+				numFailing -= 1
+			}
+
+			// Tests are allowed to flake due to transaction retry errors.
 			txnRetryErrCount := strings.Count(rawResults, "restart transaction")
-			if strings.Contains(rawResults, "1 failing") && txnRetryErrCount == 1 {
-				err = nil
-			} else if strings.Contains(rawResults, "2 failing") && txnRetryErrCount == 2 {
+			if numFailing == txnRetryErrCount {
 				err = nil
 			}
+
 			if err != nil {
 				t.Fatal(err)
 			}

